sotodlib/hwp/g3thwp.py

```diff
@@ -621,7 +629,140 @@ def write_solution(self, solved, output=None):
+        aman.wrap_new('eval', shape=('samps', ))
+
+        aman.timestamps = tod.timestamps
+        aman.stable = np.zeros(len(tod.timestamps))
+        aman.locked = np.zeros(len(tod.timestamps))
+        aman.hwp_rate = np.zeros(len(tod.timestamps))
+        aman.hwp_angle_ver1 = np.zeros(len(tod.timestamps))
+        aman.hwp_angle_ver2 = np.zeros(len(tod.timestamps))
+        aman.eval = np.zeros(len(tod.timestamps))
+        aman.save(output, h5_address, overwrite=True)
+        
+        return
+
+    
+    def write_solution_h5(self, tod, output=None, h5_address=None):
+        """
+        Output HWP angle + flags as AxisManager format
+
+        Args
+        -----
+        data: g3 file
+          data = G3tHWP.load_data(start, end)
+        output: str or None
+          output path + file name, overwrite config file
+
+        Notes
+        -----------
+        Output file format \n
+
+        - timestamp:
+            SMuRF synched timestamp
+        - hwp_angle_ver1: float
+            SMuRF synched HWP angle (calculated from raw encoder signals) in radian
+        - hwp_angle_ver2: float
+            SMuRF synched HWP angle after the template subtraction (the systematics from the non-uniform encoder slot pattern is subtracted ) in radian. \n
+            if 'eval' is zero, template subtraction is not completed and its value is same as 'hwp_angle_ver1'.
+        - slow_time: timestamp
+            time list of slow block (stable, locked. hwp_rate)
+        - stable: bool
+            if non-zero, indicates the HWP spin state is known. \n
+            i.e. it is either spinning at a measurable rate, or stationary. \n
+            When this flag is non-zero, the hwp_rate field can be taken at face value. \n
+        - locked: bool
+            if non-zero, indicates the HWP is spinning and the position solution is working. \n
+            In this case one should find the hwp_angle populated in the fast data block. \n
+        - hwp_rate: float
+            the "approximate" HWP spin rate, with sign, in revs / second. \n
+            Use placeholder value of 0 for cases when not "locked".
+        - eval: bool
+            if non-zero, the template subtraction is completed.
+        """
+        if self._output is None and output is None:
+            logger.warning('Output file not specified')
+            return
+        if output is not None:
+            self._output = output
+            
+        if len(tod.timestamps) == 0:
+            logger.warning('Input data is empty.')
+            return 
+        
+        start = int(tod.timestamps[0])-self._margin
+        end = int(tod.timestamps[-1])+self._margin
+        try:
+            data = self.load_data(start, end)
+        except Exception as e:
+            logger.error(f"Exception '{e}' thrown while loading HWP data. The specified encoder field is missing.")
+            self._write_empty_solution_h5(tod, output, h5_address)
```

Comment:
I worry this will build us a whole lot of data files filled with just zeros. Maybe we should have another way of indicating the HWP has no solutions? Curious if @mhasself  has ideas.